server/custom_vms.go

```diff
@@ -0,0 +1,481 @@
+		if err := lc.waitForLocalClusterReady(ctx); err != nil {
+			lc.customVMRestartMu.Unlock()
+			return err
+		}
+		lc.customVMRestartMu.Unlock()
+	}
+	return nil
+}
+
+func (lc *localNetwork) addSubnetValidators(ctx context.Context, baseWallet *refreshableWallet, validatorIDs []ids.ShortID) error {
+	println()
+	color.Outf("{{green}}adding all nodes as subnet validator for each subnet{{/}}\n")
+	for vmID, vmInfo := range lc.customVMIDToInfo {
+		zap.L().Info("adding all nodes as subnet validator",
+			zap.String("vm-name", vmInfo.info.VmName),
+			zap.String("vm-id", vmID.String()),
+			zap.String("subnet-id", vmInfo.subnetID.String()),
+		)
+		for _, validatorID := range validatorIDs {
+			cctx, cancel := createDefaultCtx(ctx)
+			txID, err := baseWallet.P().IssueAddSubnetValidatorTx(
+				&platformvm.SubnetValidator{
+					Validator: platformvm.Validator{
+						NodeID: validatorID,
+
+						// reasonable delay in most/slow test environments
+						Start: uint64(time.Now().Add(time.Minute).Unix()),
+						End:   uint64(time.Now().Add(100 * time.Hour).Unix()),
+						Wght:  1000,
+					},
+					Subnet: vmInfo.subnetID,
+				},
+				common.WithContext(cctx),
+				common.WithPollFrequency(5*time.Second),
+			)
+			cancel()
+			if err != nil {
+				return err
+			}
+			zap.L().Info("added the node as a subnet validator",
+				zap.String("vm-name", vmInfo.info.VmName),
+				zap.String("vm-id", vmID.String()),
+				zap.String("subnet-id", vmInfo.subnetID.String()),
+				zap.String("node-id", validatorID.String()),
+				zap.String("tx-id", txID.String()),
+			)
+		}
+	}
+	return nil
+}
+
+func (lc *localNetwork) createBlockchains(ctx context.Context, baseWallet *refreshableWallet, testKeyAddr ids.ShortID) error {
+	println()
+	color.Outf("{{green}}creating blockchain for each custom VM{{/}}\n")
+	for vmID, vmInfo := range lc.customVMIDToInfo {
+		vmName := vmInfo.info.VmName
+		vmGenesisBytes := lc.customVMNameToGenesis[vmName]
+
+		zap.L().Info("creating blockchain tx",
+			zap.String("vm-name", vmName),
+			zap.String("vm-id", vmID.String()),
+			zap.Int("genesis-bytes", len(vmGenesisBytes)),
+		)
+		cctx, cancel := createDefaultCtx(ctx)
+		blockchainID, err := baseWallet.P().IssueCreateChainTx(
+			vmInfo.subnetID,
+			vmGenesisBytes,
+			vmID,
+			nil,
+			vmName,
+			common.WithContext(cctx),
+			common.WithPollFrequency(5*time.Second),
+		)
+		cancel()
+		if err != nil {
+			return err
+		}
+
+		vmInfo.info.BlockchainId = blockchainID.String()
+		vmInfo.blockchainID = blockchainID
+		lc.customVMIDToInfo[vmID] = vmInfo
+
+		zap.L().Info("created a new blockchain",
+			zap.String("vm-name", vmName),
+			zap.String("vm-id", vmID.String()),
+			zap.String("blockchain-id", blockchainID.String()),
+		)
+	}
+	return nil
+}
+
+// Keep everything same except "whitelisted-subnets".
+func replaceWhitelistedSubnets(s string, whitelistedSubnet string) (string, error) {
```

Comment:
this function shall be more general, probably in utils

Reply:
Moved to `utils`